packages/core/src/agent/getOutboundMessageContext.ts

```diff
@@ -0,0 +1,247 @@
+  if (!lastSentMessage) {
+    agentContext.config.logger.debug(
+      `No previous sent message in thread for outbound message ${message.id}, setting up routing`
+    )
+
+    let routing: Routing | undefined = undefined
+
+    // Extract routing from out of band record if possible
+    const oobRecordRecipientRouting = outOfBandRecord?.metadata.get(OutOfBandRecordMetadataKeys.RecipientRouting)
+    if (oobRecordRecipientRouting) {
+      routing = {
+        recipientKey: Key.fromFingerprint(oobRecordRecipientRouting.recipientKeyFingerprint),
+        routingKeys: oobRecordRecipientRouting.routingKeyFingerprints.map((fingerprint) =>
+          Key.fromFingerprint(fingerprint)
+        ),
+        endpoints: oobRecordRecipientRouting.endpoints,
+        mediatorId: oobRecordRecipientRouting.mediatorId,
+      }
+    }
+
+    if (!routing) {
+      const routingService = agentContext.dependencyManager.resolve(RoutingService)
+      routing = await routingService.getRouting(agentContext, {
+        mediatorId: outOfBandRecord?.mediatorId,
+      })
+    }
+
+    ourService = {
+      id: uuid(),
+      serviceEndpoint: routing.endpoints[0],
+      recipientKeys: [routing.recipientKey],
+      routingKeys: routing.routingKeys,
+    }
+  }
+
+  // Set the parentThreadId on the message from the oob invitation
+  if (outOfBandRecord) {
+    if (!message.thread) {
+      message.setThread({
+        parentThreadId: outOfBandRecord.outOfBandInvitation.id,
+      })
+    } else {
+      message.thread.parentThreadId = outOfBandRecord.outOfBandInvitation.id
+    }
+  }
+
+  // These errors should not happen as they will be caught by the checks above. But if there's a path missed,
+  // and to make typescript happy we add these checks.
+  if (!ourService) {
+    throw new AriesFrameworkError(
+      `Could not determine our service for connection-less exchange for message ${message.id}.`
+    )
+  }
+  if (!recipientService) {
+    throw new AriesFrameworkError(
+      `Could not determine recipient service for connection-less exchange for message ${message.id}.`
+    )
+  }
+
+  // Set the service on the message and save service decorator to record (to remember our verkey)
+  // TODO: we should store this in the OOB record, but that would be a breaking change for now.
+  // We can change this in 0.5.0
+  message.service = ServiceDecorator.fromResolvedDidCommService(ourService)
+
+  const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository)
+  await didCommMessageRepository.saveOrUpdateAgentMessage(agentContext, {
+    agentMessage: message,
+    role: DidCommMessageRole.Sender,
+    associatedRecordId: associatedRecord?.id,
+  })
+
+  return new OutboundMessageContext(message, {
+    agentContext: agentContext,
+    associatedRecord,
+    serviceParams: {
+      service: recipientService,
+      senderKey: ourService.recipientKeys[0],
```

Comment:
Is the sender key the first `recipientKey`?

Reply:
Yes we always use the first key. Not really a case where we have multiple sender keys